docs/blog/2024-08-27-challenges-of-microservices.md

```diff
@@ -0,0 +1,152 @@
+
+However, in real life such outcomes are not acceptable. This means that every single player in the system must:
+
+1. Have built-in reliability mechanisms.
+2. Understand the preceding and subsequent steps of the business process to route traffic.
+
+## The Challenges
+
+So far, we have established that there are many players involved in a business process, yet there's no one player in charge of ensuring that the trains run on time. This yields three problems:
+
+1. **Reliability** in the face of infrastructure failures.
+2. **Observability** to enable system optimization and debugging.
+3. **Coupling** of microservices to each other makes it hard to modify the system in response to new business requirements.
+
+### Reliability and Correctness
+
+Processing orders is a mission-critical use-case. This means that orders should always complete and never be dropped (for example, we shouldn ever charge the customer's credit card and not ship the product to them).
+
+However, asynchronous processing such as that which I outlined above is prone to failures. For example, if you chain microservices together with direct RPC calls, a single network partition can cause an order to get stuck. Even with a reliable message broker such as Apache Kafka or AWS SQS sitting between your microservices, a write to the message broker could fail _after_ the payment went through, still resulting in a stuck order.
+
+Just as communication _between_ microservices can fail, the actions performed _by_each microservice can also fail. If the Stripe API is down, or if the credit card is invalid, we can't just stop processing the order there! We must notify the customer of what went wrong and also release the inventory that we reserved.
+
+This means that microservice developers spend countless hours building out infrastructure to support:
+* Retries
+* Dead-Letter Queues
+* Rate-limiting
+* Timeouts
+* Transactional Outbox pattern
+* SAGA Pattern
+
+Back to the domino analogy, if one domino misses the next, the entire chain just stops.
+
+### Observability
+
+The second problem with microservices is that once a process instance has started (i.e. the dominoes are falling), it is very difficult to observe what happens between steps 2 through 10. This means that multi-step processes with performance issues are hard to optimize, as there are many microservices which could be the bottleneck and it's hard to know which. Even worse, when a customer complains about a "stuck order," it is difficult to find the point of failure.
+
+As a result, microservice engineers spend time and money:
+* Slogging through logs on DataDog
+* Implementing complex distributed tracing such as Zipkin, Jaeger, or Kiali
+* Saving the state of each process instance (in our case, the `order`) in a DB just for visibility purposes at every step.
+
```

Comment:
Add: 
* Coordinating with other teams to manually understand and debug workflows